### DISCUSSION TITLE
DuxCommerce - a free Orchard Core commerce module

### DISCUSSION THREAD

**Discussion by duxonline** [OrchardCMS/OrchardCore#16548] [Show and tell]
Today I'm happy to introduce DuxCommerce - a complete Orchard Core ecommerce solution. If you are interested in the module, you can download its source code [here](https://github.com/DuxCommerce/DuxCommerce.Source). Thank you.

- **Product Listing**

![ProductListing](https://github.com/user-attachments/assets/8a189b7f-0880-4356-a51d-4c15da15492f)

- **Shopping Cart**

![ShoppingCart](https://github.com/user-attachments/assets/ed6078c3-c0f9-420f-8588-fece23d885b0)

- **Checkout**

![Checkout](https://github.com/user-attachments/assets/31bbb219-f8da-4039-90e6-234760c593cb)

- **Order Manager**

![OrderManager](https://github.com/user-attachments/assets/0cc68f08-295d-41b5-a837-820a49357828)

- **Your Account**

![YourAccount](https://github.com/user-attachments/assets/7e4b59a3-6116-429a-bcfb-9428a458806d)

- **Product Variants**

![ProductVariants](https://github.com/user-attachments/assets/fede316e-d4c5-475d-9559-7708a65809a4)

- **Coupons**

![Coupons](https://github.com/user-attachments/assets/0560fb40-8cf2-4f89-a6fc-80655be0685b)

- **Promotions**

![Promotions](https://github.com/user-attachments/assets/d636a783-a451-4524-83dc-466b52e4d13e)

**Comment by giannik** [comment]
will it be open sourced?

**Reply by duxonline** [reply]
> will it be open sourced?

It will be free and I am planning to share C# code with people who are interested in this project. I haven't decided on licence yet. I am interested to hear what the community has to say. So everyting is open to discussion at this stage.

**Reply by rwxzig** [reply]
I am interested, so let us know where to subscribe. I started a similar project too because I found the ecommerce project a bit complicated for me, so I wanted to learn by doing something.

**Comment by Piedone** [comment]
This is very nice, though any particular reason why you didn't instead contribute what you feel missing to [Orchard Core Commerce](https://github.com/OrchardCMS/OrchardCore.Commerce)?

**Comment by duxonline** [comment]
@Piedone first thank you for dojo course. When I was comparing OrchardCore with Umbraco, your course made a difference. Without it, nothing would happen or I might go with Umbraco. Actually when I was watching your videos, I was dreaming about working for Lombiq some day. 😄 That's how impressed I was by the work you guys had done.

There are a few reasons that I started this project instead of contributing to OrchardCore Commerce module.  I highlight a couple of them here.

When people ask me for advice on eCommerce solution, I always recommend Shopify because of its ease of use. The ideal solution I want to build is to hide all technical concepts like content fields, content parts, content types, etc, from end users. For example, if a user wants to enable stock tracking for a product, he should be able to do that by ticking a check box. He doesn't have to know product type or an inventory part and how to add inventory part to product type. I think this is a very fundamental difference.

The second reason has to do with coding styles or taste. When I tried to understand Commerce module code, I was struggling because I felt I had to deal with two things at the same time, content management concepts and ecommerce logic. Ecommerce itself is complex enough and adding content management to it frustrated me. A lot of times its code was just over my head. With DuxCommerce, my goal is to deal with one of them at any given time.

There are other reasons as well, but I leave them for another day.

**Reply by Piedone** [reply]
I'm glad to hear our videos helped! Working with Lombiq can still come true :D.

I see, different priorities. Though, I think if you build an e-commerce solution on Orchard Core, in the end, the content parts, fields, and others, have to be there (also because then you can allow people to leverage those Orchard features to build on top of what you provide out of the box). You can, however, build simplifications on top of them for the usual path, at least with e.g. recipes.

**Reply by duxonline** [reply]
I do use content parts and content types (for products and categories) in DuxCommerce, but not content fields. This is due to historical  reason - much backend logic already existed before migrating to Orchard Core and I was reluctant to change it. This what my content part looks like:
```
public class ProductPart : DuxPart
{
    public sealed override IRow Row { get; set; } = new ProductRow();
}
```

```
public abstract class DuxPart : ContentPart
{
    public abstract IRow Row { get; set; }
}
```
`ProductRow` is just a POCO. It is not dependent on Orchard Core. My content parts are just wrappers around POCOs, which I wrote before I run into Orchard Core. 

This approach minimizes changes I have to make. Now my backend logic is free of any Orchard Core concepts.

**Reply by Piedone** [reply]
I see, so you added preexisting code to Orchard Core.

**Reply by duxonline** [reply]
Yes.  There are obviously pros and cons to the approach. The biggest benefit for me is the clear separation of concerns - my backend contains eCommerce logic only, while UI layer focuses on CMS side, using Drivers and Shapes when needed.

There may be other people who would like to run their applications on Orchard Core. Rewriting them from scratch can be very costly or in some cases not realistic. That's why I think this hybrid approach is worth further exploring. I feel like there is not much discussion on this topic.

**Reply by Piedone** [reply]
I suppose because there are not many insights to share: if you just want to put Orchard Core into your existing ASP.NET Core app without leveraging OC in your existing code, you can just do that with minimal changes. If you do want to leverage OC features, then the work gets harder.

But bringing up case studies like this is definitely useful. Would you like to demo it on the Tuesday meeting? https://docs.orchardcore.net/en/latest/resources/meeting/

**Reply by duxonline** [reply]
Would love to. I have been pretty busy recently, but can make it happen when things quiet down

**Reply by effapp** [reply]
Hi @duxonline ,

Can you provide Product management demo, please?

Can you debate how to achive "hide all technical concepts like content fields, content parts, content types, etc, from end users" in your solution?

**Reply by hishamco** [reply]
What exactly you are looking for @effapp I might help you on this, or you can check Orchard Core Commerce repo

**Reply by effapp** [reply]
Hi @hishamco ,

I just want to refer to the implementation ideas.

**Reply by hishamco** [reply]
I thought you are looking for something that we can help you with

**Reply by duxonline** [reply]
@effapp I will. Thank you

**Comment by hishamco** [comment]
I appreciate your efforts @duxonline let us know when you open source the project, moreover, I suggest to link it in Orchard Core Commerce and contribute to that project as @Piedone suggested unless there are some reasons to driver your project by your own of the community

**Comment by giannik** [comment]
@duxonline  are there plans to open source this ? I like the UI and the simplicity.

**Reply by duxonline** [reply]
@giannik I'm still working on variant images: https://github.com/orgs/DuxCommerce/projects/4. I haven't made much progress recently due to other commitments.

I am planning to share C# code with people who are interested in the project. Please register your interest here: https://github.com/DuxCommerce/DuxCommerce.Community/discussions

**Reply by giannik** [reply]
done

**Reply by duxonline** [reply]
@giannik @effapp I am testing the commerce module on Orchard Core 2.1.3 and will share C# source code with you in the next few days once the testing is finished.

**Reply by giannik** [reply]
can help with testing too.

**Reply by effapp** [reply]
Thanks.

**Reply by cdsdev003** [reply]
@duxonline  Please invite me .
I am interested in the project

**Reply by duxonline** [reply]
@cdsdev003 This is not required any more. Please download the source code: https://github.com/DuxCommerce/DuxCommerce.Source

**Comment by duxonline** [comment]
@giannik @effapp @rwxzig @mattdunndc I have sent an invite to you guys. You should have access to the source code after accepting it.

**Reply by hishamco** [reply]
Is it OSS or private access?

**Reply by duxonline** [reply]
It's invite only. Thank you.

**Reply by hishamco** [reply]
Please invite me

I was suggesting to open source the module to get the feedback from others, and open the possibility for contributions. Otherwise Orchard Commerce might be a suitable option for all the folks

**Reply by duxonline** [reply]
Thank you @hishamco. Invite sent as requested.

**Comment by duxonline** [comment]
A Getting Started guide can be found [here](https://github.com/DuxCommerce/DuxCommerce.Community/blob/main/docs/getting-started/index.md).

**Reply by hishamco** [reply]
You need to create a recipe to populate a dummy data

**Reply by duxonline** [reply]
@hishamco That can be done. In the meantime, I am planning to create a demo store that people can download and run on their local.

**Comment by duxonline** [comment]
After some discussions with different community members, we have decided to release the source code of DuxCommerce storefront under MIT license. The source code can be downloaded [here](https://github.com/DuxCommerce/DuxCommerce.Source). 

A demo store is also included in the download. Any feedback is welcome. Thank you.

**Comment by duxonline** [comment]
This question has been asked multiple times by community. Let's answer it for all. 

Our goal for this project is to build a feature-rich, user-friendly and highly extensible shopping cart system, **making Orchard Core a competitive eCommerce platform**.

**Comment by duxonline** [comment]
DuxCommerce has been upgraded to OrchardCore 2.1.5. The project is under active development. If you think it may be useful for your future projects, please consider starring and watching the repository.

**Comment by duxonline** [comment]
DuxCommerce UI was built by backend engineers, and we’d love insights from front-end developers, web designers, and UX experts. If you have any suggestions for improvement, we’d greatly appreciate your feedback.

**Comment by duxonline** [comment]
DuxCommerce now displays all products associated with a shared option: https://github.com/DuxCommerce/DuxCommerce.Source/discussions/89

**Comment by pongpichat2** [comment]
I am currently using the  DuxCommerce and encountering an issue when trying to run the application in debug mode. When I attempt to debug the application, I receive the following error:

System.TypeInitializationException: The type initializer for '<Module>' threw an exception.
 ---> System.TypeInitializationException: The type initializer for '<Module>{063d151a-8f9b-4314-992e-ef41df52a055}' threw an exception.
 ---> System.Exception: Debugger Detected

Is this an intentional limitation of the commercial version? If so, are there any developer-specific versions or tools that would allow me to debug the application when needed? 

![image](https://github.com/user-attachments/assets/aacf926a-389e-4a96-80dc-82e31a34f02a)

**Reply by duxonline** [reply]
@pongpichat2 This is a known issue and will be addressed in the next release.

**Reply by hishamco** [reply]
IMHO, all the detailed discussion should go to the DuxCommerce repo. This discussion intends to show and tell the community about `DuxCommerce` and its features
